File: src/domain/entities/Recipe.ts
import { EmptyRecipeException } from '../errors/recipe/EmptyRecipeException.js'
import { InvalidQuantityException } from '../errors/recipe/InvalidQuantityException.js'
import type { Id } from '../valueobjects/Id.js'
import type { Money } from '../valueobjects/Money.js'
import { RecipeIngredient } from '../valueobjects/RecipeIngredient.js'

export class Recipe {
  private readonly id: Id
  private name: string
  private ingredients: RecipeIngredient[]
  private totalCost: Money

  constructor(id: Id, name: string, ingredients: RecipeIngredient[], totalCost: Money) {
    this.validateName(name)
    this.validateIngredients(ingredients)

    this.id = id
    this.name = name.trim()
    this.ingredients = [...ingredients] // Defensive copy
    this.totalCost = totalCost
  }

  private validateName(name: string): void {
    if (!name || name.trim().length === 0) {
      throw new Error('Recipe name cannot be empty')
    }

    if (name.trim().length > 255) {
      throw new Error('Recipe name too long (max 255 characters)')
    }
  }

  private validateIngredients(ingredients: RecipeIngredient[]): void {
    if (!ingredients || ingredients.length === 0) {
      throw new EmptyRecipeException()
    }

    // Check for duplicate ingredients
    const ingredientIds = new Set<string>()
    for (const ingredient of ingredients) {
      const idValue = ingredient.getIngredientId().getValue()
      if (ingredientIds.has(idValue)) {
        throw new Error(`Duplicate ingredient found: ${idValue}`)
      }
      ingredientIds.add(idValue)

      // Validate quantity
      if (ingredient.getQuantity().isNegative() || ingredient.getQuantity().isZero()) {
        throw new InvalidQuantityException(ingredient.getQuantity().toString())
      }
    }
  }

  getId(): Id {
    return this.id
  }

  getName(): string {
    return this.name
  }

  getIngredients(): readonly RecipeIngredient[] {
    return [...this.ingredients] // Return defensive copy
  }

  getTotalCost(): Money {
    return this.totalCost
  }

  updateName(newName: string): void {
    this.validateName(newName)
    this.name = newName.trim()
  }

  addIngredient(ingredient: RecipeIngredient): void {
    // Check if ingredient already exists
    const existingIngredient = this.ingredients.find((ing) =>
      ing.getIngredientId().equals(ingredient.getIngredientId())
    )

    if (existingIngredient) {
      throw new Error(
        `Ingredient ${ingredient.getIngredientId().getValue()} already exists in recipe`
      )
    }

    this.ingredients.push(ingredient)
  }

  removeIngredient(ingredientId: Id): void {
    const index = this.ingredients.findIndex((ing) => ing.getIngredientId().equals(ingredientId))

    if (index === -1) {
      throw new Error(`Ingredient ${ingredientId.getValue()} not found in recipe`)
    }

    this.ingredients.splice(index, 1)

    if (this.ingredients.length === 0) {
      throw new EmptyRecipeException()
    }
  }

  updateIngredientQuantity(ingredientId: Id, newQuantity: string | number): void {
    const ingredient = this.ingredients.find((ing) => ing.getIngredientId().equals(ingredientId))

    if (!ingredient) {
      throw new Error(`Ingredient ${ingredientId.getValue()} not found in recipe`)
    }

    // Create new ingredient with updated quantity
    const updatedIngredient = new RecipeIngredient(
      ingredient.getIngredientId(),
      newQuantity,
      ingredient.getUnit()
    )

    // Replace the ingredient
    const index = this.ingredients.findIndex((ing) => ing.getIngredientId().equals(ingredientId))
    this.ingredients[index] = updatedIngredient
  }

  updateTotalCost(newTotalCost: Money): void {
    this.totalCost = newTotalCost
  }

  getIngredientCount(): number {
    return this.ingredients.length
  }

  hasIngredient(ingredientId: Id): boolean {
    return this.ingredients.some((ing) => ing.getIngredientId().equals(ingredientId))
  }

  equals(other: Recipe): boolean {
    if (!(other instanceof Recipe)) {
      return false
    }
    return this.id.equals(other.id)
  }

  toString(): string {
    return `Recipe{id=${this.id.toString()}, name='${this.name}', ingredients=${this.ingredients.length}, totalCost=${this.totalCost.toString()}}`
  }
}

File: api/src/main/java/br/unifor/costify/domain/entity/Recipe.java
package br.unifor.costify.domain.entity;

import br.unifor.costify.domain.contracts.IdGenerator;
import br.unifor.costify.domain.errors.recipe.EmptyRecipeException;
import br.unifor.costify.domain.errors.recipe.InvalidTotalCostException;
import br.unifor.costify.domain.valueobject.Id;
import br.unifor.costify.domain.valueobject.Money;
import br.unifor.costify.domain.valueobject.RecipeIngredient;
import java.util.ArrayList;
import java.util.List;

public class Recipe {
  private Id id;
  private String name;
  private List<RecipeIngredient> ingredients;
  private Money totalCost;


  public Recipe(Id id, String name, List<RecipeIngredient> ingredients, Money totalCost) {
    this.validate(name, ingredients);
    if (totalCost == null) {
      throw new InvalidTotalCostException("Total cost cannot be null");
    }
    this.id = id;
    this.name = name;
    this.ingredients = new ArrayList<>(ingredients);
    this.totalCost = totalCost;
  }


  public Recipe(IdGenerator idGenerator, String name, List<RecipeIngredient> ingredients, Money totalCost) {
    this.validate(name, ingredients);
    if (totalCost == null) {
      throw new InvalidTotalCostException("Total cost cannot be null");
    }
    this.id = Id.generate(idGenerator);
    this.name = name;
    this.ingredients = new ArrayList<>(ingredients);
    this.totalCost = totalCost;
  }

  public Id getId() {
    return id;
  }

  public String getName() {
    return name;
  }

  public List<RecipeIngredient> getIngredients() {
    return new ArrayList<>(ingredients);
  }

  public Money getTotalCost() {
    return totalCost;
  }


  public void addIngredient(RecipeIngredient ingredient) {
    this.ingredients.add(ingredient);
  }

  public void removeIngredient(Id ingredientId) {
    List<RecipeIngredient> updatedIngredients =
        ingredients.stream().filter(ri -> !ri.getIngredientId().equals(ingredientId)).toList();

    if (updatedIngredients.isEmpty()) {
      throw new EmptyRecipeException("Recipe must have at least one ingredient");
    }

    this.ingredients = new ArrayList<>(updatedIngredients);
  }

  public void updateTotalCost(Money newTotalCost) {
    if (newTotalCost == null) {
      throw new InvalidTotalCostException("Total cost cannot be null");
    }
    this.totalCost = newTotalCost;
  }

  public void updateName(String newName) {
    if (newName == null || newName.isBlank()) {
      throw new IllegalArgumentException("Recipe name cannot be null or empty");
    }
    this.name = newName;
  }

  public void updateIngredients(List<RecipeIngredient> newIngredients) {
    if (newIngredients == null || newIngredients.isEmpty()) {
      throw new EmptyRecipeException("Recipe must have at least one ingredient");
    }
    this.ingredients = new ArrayList<>(newIngredients);
  }

  private void validate(String name, List<RecipeIngredient> ingredients) {
    if (ingredients == null || ingredients.isEmpty()) {
      throw new EmptyRecipeException("Recipe must have at least one ingredient");
    }
  }
}
